simplified-ui-catalog/src/main/java/org/nypl/simplified/ui/catalog/AutoClearedValue.kt

```diff
@@ -0,0 +1,47 @@
+package org.nypl.simplified.ui.catalog
+
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+/**
+ * https://github.com/android/architecture-components-samples/blob/master/GithubBrowserSample/app/src/main/java/com/android/example/github/util/AutoClearedValue.kt
+ *
+ * A lazy property that gets cleaned up when the fragment's view is destroyed.
+ *
+ * Accessing this variable while the fragment's view is destroyed will throw NPE.
+ */
+class AutoClearedValue<T : Any>(val fragment: Fragment) : ReadWriteProperty<Fragment, T> {
+  private var _value: T? = null
+
+  init {
+    fragment.lifecycle.addObserver(object : DefaultLifecycleObserver {
+      override fun onCreate(owner: LifecycleOwner) {
+        fragment.viewLifecycleOwnerLiveData.observe(fragment) { viewLifecycleOwner ->
+          viewLifecycleOwner?.lifecycle?.addObserver(object : DefaultLifecycleObserver {
+            override fun onDestroy(owner: LifecycleOwner) {
+              _value = null
+            }
+          })
+        }
+      }
+    })
+  }
+
+  override fun getValue(thisRef: Fragment, property: KProperty<*>): T {
+    return _value ?: throw IllegalStateException(
+      "should never call auto-cleared-value get when it might not be available"
+    )
+  }
+
+  override fun setValue(thisRef: Fragment, property: KProperty<*>, value: T) {
+    _value = value
+  }
+}
+
+/**
+ * Creates an [AutoClearedValue] associated with this fragment.
+ */
+fun <T : Any> Fragment.autoCleared() = AutoClearedValue<T>(this)
```

Comment:
I dislike the use of the view's lifecycle being implicit here. We could name this `viewAutoCleared` or add a boolean argument `viewLifecycle` and make it working with the fragment's lifecycle too.

Reply:
I also wasn't a huge fan of the name `autoCleared` so I'm happy to change it to better reflect what it is actually doing. 

I'm thinking `viewLifecycleAware()` or something like that.

Reply:
Agreed.